internal/provider/resource_apply_module.go

```diff
@@ -0,0 +1,564 @@
+		model: &applyModule,
+	}, &didRun)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Update the plan with the computed ID.
+	applyModule.ID = types.StringValue(uuid.New().String())
+	applyModule.ModuleVersion = types.StringValue(didRun.moduleVersion)
+
+	// Set the response state to the fully-populated plan, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, applyModule)...)
+}
+
+func (t *applyModuleResource) Read(ctx context.Context,
+	req resource.ReadRequest, resp *resource.ReadResponse) {
+
+	// Get the current state.
+	var state ApplyModuleModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var appliedModuleInfo appliedModuleInfo
+	resp.Diagnostics.Append(t.getCurrentApplied(ctx, state, &appliedModuleInfo)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Update the state with the computed module version.
+	if appliedModuleInfo.moduleVersion != nil {
+		state.ModuleVersion = types.StringValue(*appliedModuleInfo.moduleVersion)
+	}
+
+	// TODO: Eventually, when the API and SDK support speculative plan runs with a module source,
+	// this should do a speculative plan run here to determine whether changes are needed.
+
+	// Set the refreshed state, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (t *applyModuleResource) Update(ctx context.Context,
+	req resource.UpdateRequest, resp *resource.UpdateResponse) {
+
+	// Retrieve values from plan.
+	var plan ApplyModuleModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// TODO: Please note that when the API and SDK support speculative runs with a module source,
+	// this will need to look at the results from the Read method's speculative run to determine
+	// whether to do an update.  A way will have to be found to force Terraform to allow the update.
+
+	// Do the run.
+	var didRun doRunOutput
+	resp.Diagnostics.Append(t.doRun(ctx, &doRunInput{
+		model: &plan,
+	}, &didRun)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Capture the module version in case it changed.
+	plan.ModuleVersion = types.StringValue(didRun.moduleVersion)
+
+	// Set the response state to the fully-populated plan, with or without error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+}
+
+func (t *applyModuleResource) Delete(ctx context.Context,
+	req resource.DeleteRequest, resp *resource.DeleteResponse) {
+
+	// Get the current state.
+	var state ApplyModuleModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var appliedModuleInfo appliedModuleInfo
+	resp.Diagnostics.Append(t.getCurrentApplied(ctx, state, &appliedModuleInfo)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// If the module source or module version if available and differs, error out.
+	if appliedModuleInfo.moduleSource != nil {
```

Comment:
This should return an error if the module source is nill but the workspace has a current state since that'll indicate the workspace source differs from what is currently deployed.

Reply:
It doesn't look like this is handling the case where the workspace state is empty in which case it shouldn't return an error since the state has already been destroyed.